src/TLE.ts

```diff
@@ -874,65 +874,95 @@ export class Parser {
         assert(1 <= stringValue.length, "TLE strings must be at least 1 character.");
         assert(Utilities.isQuoteCharacter(stringValue[0]), "The first character in a TLE string must be a quote character.");
 
-        let leftSquareBracketToken: Token = null;
+        const leftSquareBracketToken: Token = null;
         let expression: Value = null;
-        let rightSquareBracketToken: Token = null;
-        let errors: language.Issue[] = [];
+        const rightSquareBracketToken: Token = null;
+        const errors: language.Issue[] = [];
+        let result: { stringValue: string; leftSquareBracketToken: Token; expression: Value; rightSquareBracketToken: Token; errors: language.Issue[] } = { stringValue, leftSquareBracketToken, expression, rightSquareBracketToken, errors };
 
         if (3 <= stringValue.length && stringValue.substr(1, 2) === "[[") {
```

Comment:
These changes seem more complex than necessary.  Also, note that this code was ported from C# - the checks against the string length aren't necessary, out of bounds string indices just return undefined.

I'd think the following single change would do it:

        if (
            stringValue[1] !== "[" // Does not start with '['
            || stringValue.substr(1, 2) === "[[" // Starts with '[['
            || (stringValue[1] === '[' && stringValue[stringValue.length - 1] === ']') // Starts with '[' but doesn't end with ']'
        ) {
            // This is considered a plain string
            expression = new StringValue(Token.createQuotedString(0, stringValue));
        } else {
          // This is considered an expression

With that first check I added to verify it starts with '[', you should also be able to get rid of the following if/else (this will keep us from tokenizing the entire string when it doesn't even start with '[' - I didn't write this code :-)).


            if (!tokenizer.hasCurrent() || tokenizer.current.getType() !== TokenType.LeftSquareBracket) {
                expression = new StringValue(Token.createQuotedString(0, stringValue));
            } else {

Reply:
Note: This exposes some incorrect testcases - currently since the string is tokenized before checking for [, whitespace is thrown away, and it considers strings like " [expr()] " to be an expression, but when I run that through the portal, it considers it a string, not an expression.